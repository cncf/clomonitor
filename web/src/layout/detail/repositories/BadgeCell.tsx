import classNames from 'classnames';
import { isUndefined } from 'lodash';

import getCategoryColor from '../../../utils/getCategoryColor';
import styles from './BadgeCell.module.css';

interface Props {
  value?: number;
  cellClassName?: string;
  onClick: () => void;
}

const BadgeCell = (props: Props) => {
  if (isUndefined(props.value))
    return (
      <td className={props.cellClassName || 'd-none d-md-table-cell align-middle'}>
        <div className="mx-auto px-2 text-center">
          <span className="text-muted">n/a</span>
        </div>
      </td>
    );

  const color = getCategoryColor(props.value);

  return (
    <td
      className={classNames(
        styles.hoverableCell,
        { 'd-none d-md-table-cell align-middle': isUndefined(props.cellClassName) },
        props.cellClassName
      )}
      onClick={props.onClick}
      role="button"
    >
      <div
        className={`mx-auto px-2 text-center text-dark fw-bold ${styles.badge}`}
        style={{
          borderBottomColor: `var(--rm-${color})`,
        }}
      >
        {props.value}
      </div>
    </td>
  );
};

export default BadgeCell;
