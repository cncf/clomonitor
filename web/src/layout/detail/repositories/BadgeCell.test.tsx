import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import BadgeCell from './BadgeCell';

const mockOnClick = jest.fn();

const defaultProps = {
  onClick: mockOnClick,
};

describe('Badge', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(
      <table>
        <tbody>
          <tr>
            <BadgeCell {...defaultProps} value={80} />
          </tr>
        </tbody>
      </table>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders badge', () => {
      render(
        <table>
          <tbody>
            <tr>
              <BadgeCell {...defaultProps} value={80} />
            </tr>
          </tbody>
        </table>
      );

      expect(screen.getByText('80')).toBeInTheDocument();
    });

    it('renders badge with undefined value', () => {
      render(
        <table>
          <tbody>
            <tr>
              <BadgeCell {...defaultProps} />
            </tr>
          </tbody>
        </table>
      );

      expect(screen.getByText('n/a')).toBeInTheDocument();
      expect(screen.queryByRole('button')).toBeNull();
    });

    it('clicks on anchor', () => {
      render(
        <table>
          <tbody>
            <tr>
              <BadgeCell {...defaultProps} value={70} />
            </tr>
          </tbody>
        </table>
      );

      const btn = screen.getByRole('button');
      userEvent.click(btn);

      expect(mockOnClick).toHaveBeenCalledTimes(1);
    });
  });
});
