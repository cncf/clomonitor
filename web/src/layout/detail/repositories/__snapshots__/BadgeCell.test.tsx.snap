// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge creates snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="hoverableCell d-none d-md-table-cell align-middle"
          role="button"
        >
          <div
            class="mx-auto px-2 text-center text-dark fw-bold badge"
          >
            80
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
